Userland/Libraries/LibTest/Randomized/Generator.h

```diff
@@ -0,0 +1,321 @@
+/*
+ * Copyright (c) 2023, Martin Janiczek <martin@janiczek.cz>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <LibTest/Macros.h>
+#include <LibTest/Randomized/RandomRun.h>
+
+#include <AK/Function.h>
+#include <AK/Random.h>
+#include <AK/String.h>
+#include <AK/StringView.h>
+#include <AK/Tuple.h>
+
+namespace Test {
+namespace Randomized {
+
+// Returns a random double value in range 0..1.
+inline double get_random_probability()
+{
+    static constexpr u32 max_u32 = NumericLimits<u32>::max();
+    u32 random_u32 = AK::get_random_uniform(max_u32);
+    return static_cast<double>(random_u32) / static_cast<double>(max_u32);
+}
+
+// Generators take random bits from the RandomnessSource and return a value
+// back.
+//
+// Example:
+// - Gen::u32(5,10) --> 9, 7, 5, 10, 8, ...
+namespace Gen {
+
+// An unsigned integer generator.
+//
+// The minimum value will always be 0.
+// The maximum value is given by user in the argument.
+//
+// Gen::unsigned_int(10) -> value 5, RandomRun [5]
+//                       -> value 8, RandomRun [8]
+//                       etc.
+//
+// Shrinks towards 0.
+inline u32 unsigned_int(u32 max)
+{
+    u32 random = Test::randomness_source().draw_value(max, [&]() {
+        return AK::get_random_uniform(max + 1);
+    });
+    return random;
+}
+
+// An unsigned integer generator in a particular range.
+//
+// Gen::unsigned_int(3,10) -> value 3,  RandomRun [0]
+//                         -> value 8,  RandomRun [5]
+//                         -> value 10, RandomRun [7]
+//                         etc.
+//
+// In case `min == max`, the RandomRun footprint will be smaller, as we'll
+// switch to a `constant` and won't need any randomness to generate that
+// value:
+//
+// Gen::unsigned_int(3,3) -> value 3, RandomRun [] (always)
+//
+// Shrinks towards the smaller argument.
+inline u32 unsigned_int(u32 min, u32 max)
+{
+    VERIFY(max >= min);
+    if (min == max) {
+        return min;
+    }
```

Comment:
I guess this could also be a `max == 0` case in `unsigned_int` already?

Reply:
I don't currently see how this would make things any different:
```diff
-if (min == max) {
+if (min == max || max == 0) {
     return min;
 }
```
given we already have the `VERIFY(max >= min);` case one line above this one. (`max >= min && max == 0` is the same as `max >= min && min == max`)
Am I misunderstanding your intended change?

Reply:
Ah, whoops, that `unsigned_int` was ambiguous.

I meant that calling `unsigned_int(min, max)` with `min == max` essentially amounts to `unsigned_int(0) + min`, and the `unsigned_int(0)` part we presumably also don't want to use randomness for.

What I'm proposing is this:

```diff
diff --git a/Userland/Libraries/LibTest/Randomized/Generator.h b/Userland/Libraries/LibTest/Randomized/Generator.h
index c65ceeb624..62d23c9a6a 100644
--- a/Userland/Libraries/LibTest/Randomized/Generator.h
+++ b/Userland/Libraries/LibTest/Randomized/Generator.h
@@ -45,6 +45,9 @@ namespace Gen {
 // Shrinks towards 0.
 inline u32 unsigned_int(u32 max)
 {
+    if (max == 0)
+        return 0;
+
     u32 random = Test::randomness_source().draw_value(max, [&]() {
         return AK::get_random_uniform(max + 1);
     });
@@ -68,9 +71,6 @@ inline u32 unsigned_int(u32 max)
 inline u32 unsigned_int(u32 min, u32 max)
 {
     VERIFY(max >= min);
-    if (min == max) {
-        return min;
-    }
     return unsigned_int(max - min) + min;
 }
 
```

Reply:
That makes sense. Will do, thank you!